js/about/preferences.js

```diff
@@ -161,19 +163,73 @@ class GeneralTab extends ImmutableComponent {
   }
 }
 
+class SearchSelectEntry extends ImmutableComponent {
+  shouldComponentUpdate (nextProps, nextState) {
+    return this.props.settings.get('search.default-search-engine') !== nextProps.settings.get('search.default-search-engine')
+  }
+  render () {
+    return <div>
+    {this.props.settings.get('search.default-search-engine') === this.props.name ? 'x' : ''}
+    </div>
+  }
+}
+
+class SearchEntry extends ImmutableComponent {
+  render () {
+    return <div>
+      <span style={this.props.iconStyle}>
+      </span>
+      <span style={this.props.textStyle}>{this.props.name}</span>
+    </div>
+  }
+}
+
+class SearchShortcutEntry extends ImmutableComponent {
+  render () {
+    return <div style={this.props.textStyle}>
+      {this.props.shortcut}
+    </div>
+  }
+}
+
 class SearchTab extends ImmutableComponent {
+  get searchProviders () {
+    let entries = searchProviders.providers
+    let array = []
+    const iconSize = 16
+    entries.forEach((entry) => {
+      let iconStyle = {
+        backgroundImage: `url(${entry.image})`,
+        minWidth: iconSize,
+        width: iconSize,
+        backgroundSize: iconSize,
+        height: iconSize,
+        display: 'inline-block',
+        verticalAlign: 'middle'
+      }
+      let textStyle = {
+        paddingLeft: '5px',
+        verticalAlign: 'middle'
+      }
+      array.push([<SearchSelectEntry name={entry.name} settings={this.props.settings} />,
+        <SearchEntry name={entry.name} iconStyle={iconStyle} textStyle={textStyle}
+          onChangeSetting={this.props.onChangeSetting} />,
+        <SearchShortcutEntry shortcut={entry.shortcut} textStyle={textStyle} />])
+    })
+    return array
+  }
+
+  hoverCallback (rows) {
+    this.props.onChangeSetting(settings.DEFAULT_SEARCH_ENGINE, rows[1].props.children.props.name)
+  }
+
   render () {
     return <div>
       <div className='sectionTitle' data-l10n-id='searchSettings' />
-      <SettingsList>
-        <SettingItem dataL10nId='defaultSearchEngine'>
-          <select value={getSetting(settings.DEFAULT_SEARCH_ENGINE, this.props.settings)}
-            onChange={changeSetting.bind(null, this.props.onChangeSetting, settings.DEFAULT_SEARCH_ENGINE)}>
-            <option value='content/search/google.xml'>Google</option>
-            <option value='content/search/duckduckgo.xml'>DuckDuckGo</option>
-          </select>
-        </SettingItem>
-      </SettingsList>
+      <SettingItem dataL10nId='defaultSearchEngine'>
+        <SortableTable headings={['Default', 'Search Engines', 'Engine Go Key (type first)']} rows={this.searchProviders}
```

Comment:
I suggest to make sortableTable take in a list of l10n ids alternately to display those.